v2/tools/mangle-test-json/main.go

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"sort"
+	"strings"
+	"time"
+)
+
+type JSONFormat struct {
+	Time    time.Time
+	Action  string
+	Package string
+	Test    string
+	Output  string
+}
+
+type TestRun struct {
+	Action  string
+	Package string
+	Test    string
+	Output  []string
+	RunTime time.Duration
+}
+
+func main() {
+	for _, testOutputFile := range os.Args[1:] {
+		fmt.Printf("# `%s`\n\n", testOutputFile)
+
+		byPackage := loadJSON(testOutputFile)
+
+		packages := []string{}
+		for k, v := range byPackage {
+			packages = append(packages, k)
+			sort.Slice(v, func(i, j int) bool {
+				return v[i].Test < v[j].Test
+			})
+		}
+
+		sort.Strings(packages)
+
+		printSummary(packages, byPackage)
+		printDetails(packages, byPackage)
+		printSlowTests(byPackage)
+	}
+}
+
+func min(i, j int) int {
+	if i <= j {
+		return i
+	}
+
+	return j
+}
+
+func actionSymbol(d TestRun) string {
+	switch d.Action {
+	case "pass":
+		return "✅"
+	case "fail":
+		return "❌"
+	case "skip":
+		return "⏭️"
+	default:
+		panic(fmt.Sprintf("unhandled action: %s", d.Action))
+	}
+}
+
+func loadJSON(testOutputFile string) map[string][]TestRun {
+	content, err := ioutil.ReadFile(testOutputFile)
+	if err != nil {
+		log.Fatalf("%e", err)
+	}
+
+	// make test output into valid JSON
+	jsonData := "[" + strings.Join(strings.Split(strings.Trim(string(content), " \n\r"), "\n"), ",") + "]"
+
+	data := []JSONFormat{}
+	err = json.Unmarshal([]byte(jsonData), &data)
+	if err != nil {
+		log.Fatalf("%e", err)
+	}
+
+	// track when each test started running
+	startTimes := make(map[string]time.Time)
+	runTimes := make(map[string]time.Duration)
+	outputs := make(map[string][]string)
+	key := func(d JSONFormat) string {
+		return d.Package + "/" + d.Test
+	}
+
+	// package → list of tests
+	byPackage := make(map[string][]TestRun, len(data))
+	for _, d := range data {
+		switch d.Action {
+		case "run":
+			if startTimes[key(d)] != (time.Time{}) {
+				panic("run while already running")
+			}
+			startTimes[key(d)] = d.Time
+		case "pause":
+			if startTimes[key(d)] == (time.Time{}) {
+				panic("pause while not running")
+			}
+			runTimes[key(d)] += d.Time.Sub(startTimes[key(d)])
+			startTimes[key(d)] = time.Time{}
+		case "cont":
+			// cont while still in running state happens sometimes (???)
+			// so don't check
+			startTimes[key(d)] = d.Time
+		case "output":
+			outputs[key(d)] = append(outputs[key(d)], d.Output)
```

Comment:
Maybe the markdown summary shouldn't include the raw output, and only a summary of tests, execution times, and outcomes? We can look at the full logs for the details.

Reply:
Have changed it so we only show failed tests and trim the output maximum of 300 lines. Also the full output is logged on stderr so it appears in the standard log (since with `-json` all output is suppressed/captured).